core/src/main/java/fr/sncf/osrd/OSRDException.java

```diff
@@ -0,0 +1,28 @@
+package fr.sncf.osrd;
+
+import com.squareup.moshi.JsonAdapter;
+import fr.sncf.osrd.utils.Reflection;
+
+public abstract class OSRDException extends RuntimeException {
+
+    private static final long serialVersionUID = 1197516372515951853L;
+
+    /** Detailed error message */
+    public final String message;
+
+    /** Whether this is an internal or user error, used to determine if we send a 400 or 500 code */
+    public final ErrorCause cause;
+
+    public enum ErrorCause {
+        INTERNAL,
+        USER
+    }
+
+    protected OSRDException(String message, ErrorCause cause) {
+        this.message = message;
+        this.cause = cause;
+    }
+
+    public static final JsonAdapter<OSRDException> adapter =
+            Reflection.makeJsonAdapterFromSubtypes(OSRDException.class, "type");
```

Comment:
I don't see the hierarchical error type identifier as discussed in https://github.com/DGEXSolutions/osrd/issues/557#issuecomment-1044311471
my understanding is that the class name is used as the error type identifier, which does not sound like a good idea, as this makes name clashes easier, and renaming the class has direct effect on the public API

Reply:
True, but then the alternative is what made you go "wtf". I assumed this was what you had in mind